data/modules/TradeShips/Trader.lua

```diff
@@ -0,0 +1,355 @@
+				starport, distance = next_starport, next_distance
+			end
+		end
+	end
+	return starport or current
+end
+
+Trader.addFuel = function (ship)
+	local drive = ship:GetEquip('engine', 1)
+
+	-- a drive must be installed
+	if not drive then
+		Core.log:add(ship, 'Ship has no drive!')
+		return nil
+	end
+
+	-- the last character of the fitted drive is the class
+	-- the fuel needed for max range is the square of the drive class
+	local count = drive.capabilities.hyperclass ^ 2
+
+	-- account for fuel it already has
+	count = count - ship:CountEquip(e.cargo.hydrogen)
+
+	local added = ship:AddEquip(e.cargo.hydrogen, count)
+
+	return added
+end
+
+-- TRADER DEFERRED TASKS
+--
+-- call the passed function in a specified time, checking whether we are still
+-- in this system
+Trader.callInThisSystem = function(t, fnc)
+	local current_path = Game.system.path
+	Timer:CallAt(t, function()
+		if Game.system and current_path:IsSameSystem(Game.system.path) then fnc() end
+	end)
+end
+
+local trader_task = {}
+-- a table of functions that can be assigned for delayed execution by the trader
+-- { ["fnc1"] = fnc1, ... }
+-- made to serialize pending job execution
+-- the task function prototype should be:
+-- function(ship) -> bool
+-- return true, if another task was assigned from the task
+-- else return false or nil or nothing
+
+trader_task.doUndock = function (ship)
+	-- the player may have left the system or the ship may have already undocked
+	if ship:exists() and ship:GetDockedWith() then
+		-- we load goods before departure
+		Trader.addCargo(ship, 'export')
+		if not ship:Undock() then
+			-- unable to undock, try again later
+			Trader.assignTask(ship, Game.time + Core.WAIT_FOR_NEXT_UNDOCK, 'doUndock')
+			return true
+		end
+	end
+end
+
+trader_task.doRedock = function(ship)
+	trader = Core.ships[ship]
+	if trader then
+		if ship:exists() and ship.flightState ~= 'HYPERSPACE' then
+			trader['starport'] = Trader.getNearestStarport(ship, trader.starport)
+			ship:AIDockWith(trader.starport)
+			trader['status'] = 'inbound'
+			trader.ts_error = "dock_aft_6h"
+		end
+	end
+end
+
+-- at the appointed time the specified function will be executed, in addition,
+-- it will be executed only in the same system in which it is assigned, and
+-- also the assignment can be serialized
+-- ship: object
+-- delay: absolute game time, when to execute the task
+-- fn_name: string, one of the keys in the table 'trader_task'
+Trader.assignTask = function(ship, delay, fn_name)
+	local trader = Core.ships[ship]
+	if trader then
+		trader.delay = delay
+		trader.fnc = fn_name
+		Trader.callInThisSystem(delay, function()
+			if(not trader_task[fn_name](ship)) then
+				-- after the function ends, remove the assignment
+				-- but we don't do this if the function returned something, because in
+				-- that case it assigned a new task
+				trader.delay = nil
+				trader.fnc = nil
+			end
```

Comment:
Instead of returning a `true` value to avoid cleanup, I'd recommend statically setting `trader.delay` etc. to nil before calling the function; if the function needs those variables as context you can pass them in from the upvalues as extra parameters. E.g.:
```lua
trader.delay = nil
trader.fnc = nil
trader_task[fn_name](ship, delay, fn_name)
```

Also it seems slightly redundant to be setting those values on the ship, as AFAICT they're only used here, as upvalues, and the copy on the ship object is there for debugging purposes.

Reply:
Done! Indeed, that is much better!